### PR TITLE
interfaces: builtin: add pluggable storage interface

### PR DESCRIPTION
**Pull request by ssweeny**
This supersedes https://github.com/snapcore/snapd/pull/1526 with some additional changes suggested by @morphis. It still needs a review from @jdstrand.

### CODE REVIEW

File: interfaces/builtin/pluggable_storage.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+/bin/umount ixr,
+
+# mount/umount (via libmount) track some mount info in these files
+/run/mount/utab wrl,
+
+# USB dev files (sda, sdb, etc...)
+/dev/sd* r,
+
+# SD card dev files
+/dev/mmcblk* r,
```
Comment (jdstrand) [inline interfaces/builtin/pluggable_storage.go]: This interface appears to intend (there is no docs/interfaces.md update and the Description is too terse so just going by the policy) to allow readonly access to pluggable devices but it actually grants ownership of the device.

The mount rules for /dev/sd* mean that the device can be mounted without mount rules, which defaults to read/write. As such, a snap with this interface connected can simply: `mount /dev/sda1 $SNAP_DATA/mnt` and on a typical install modify all files in the sda1 partition-- /etc/shadow, sshd_config, pam, grub, security policy, systemd jobs, etc, etc. On an all snaps system this interface could be used to replace security policy, systemd jobs, etc, etc. Once that is done, it is trivial to completely own the device on reboot (and a boot can be retriggered easily enough with what this interface grants). The same can be said for devices the boot off of SD cards like beaglebone or rpi due to the mount rule for /dev/mmcblk*. We could adjust the mount options in the apparmor rule to be readonly but I suspect that will not provide the required functionality that this interface intends to provide.

Furthermore, the /dev/sd* and /dev/mmcblk* read access means that this interface grants complete read access to these raw devices and partitions (and therefore the entire device) and that sensitive system, user and application data is granted to any snap with this interface connected.

The sys_admin capability is unfortunate but required due to the use of mount, and due to limitation in the Linux capabilities system, this grants more than simply mount.

In all, -1 for this interface as implemented.

This needs quite a bit of design to work well and safely. I think there are at least three options for a proper solution that would need to be considered:

1. use something along the lines of the storage framework as designed by @vosst for Ubuntu Touch
2. snapd should grow a device monitor service that is able to detect pluggable device events and make the the filesystem available to snaps that have this interface connected
3. snapd should grow a device monitor service that is able to detect pluggable device events and make them available to snaps that have this interface connected via the SecurityUDev backend. Ie, tag assigned devices so they are available to snaps.

'1' would need @vosst to comment and review for snappy (IIRC it was designed only with Touch in mind, but it may be adaptable to snapd)

'2' is a reimplementation of '1' AIUI in a snappy world and does not require giving the snap access to `mount`. Instead, the device monitor mounts and the security policy of the snap is adjusted accordingly (and perhaps we would simply use a bind mount into /var/lib/$SNAP_NAME/mounts so the security policy doesn't have to be adjusted).

'3' would still require the mount command, but we are able to limit what can be mounted to only those devices that are in the device cgroup, so on a device with /dev/sda for the SSD/hardrive, /dev/sda1 wouldn't be there (since the device monitor wouldn't add it) but /dev/sdb is. Specifically, the device monitor would update the udev rules for any plugged in removable block device for the snap, which would trigger the udev command to add the devices to the snap's device cgroup.

In summary-- we don't want the snap to have to run mount commands on random devices in /dev itself. Instead, something else (ie, a trusted helper outside of the snap's process space) is in charge of mounting the devices and making the filesystem (not the raw device) available to the snap. This needs proper design with the interfaces team along with other members of the security team (eg, @tyhicks and/or Emily) and likely @vosst so that whatever is done works with converged devices.

File: interfaces/builtin/pluggable_storage.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+# Needed for mount/unmount operations
+capability sys_admin,
+
+# Mount/unmount USB storage devices
+mount /dev/sd* -> /var/snap/${SNAP_NAME}/${SNAP_REVISION}/**,
+umount /var/snap/${SNAP_NAME}/${SNAP_REVISION}/**,
+
+# Mount/unmount SD cards
+mount /dev/mmcblk* -> /var/snap/${SNAP_NAME}/${SNAP_REVISION}/**,
+umount /var/snap/${SNAP_NAME}/${SNAP_REVISION}/**,
```
Comment (jdstrand) [inline interfaces/builtin/pluggable_storage.go]: Note, ${SNAP_NAME} is invalid AppArmor syntax. You should use @{SNAP_NAME}.

File: interfaces/builtin/pluggable_storage.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+# Allow calling the system mount/umount binaries to do the dirty work
+/bin/mount ixr,
+/bin/umount ixr,
+
+# mount/umount (via libmount) track some mount info in these files
+/run/mount/utab wrl,
```
Comment (jdstrand) [inline interfaces/builtin/pluggable_storage.go]: This is an info leak similar to mount-observe: mount arguments might contain sensitive information and may be able to enumerate what snaps are installed.

**Comment by vosst** [comment]
Please note that the storage management component taking care of mounting/formatting sd cards on Ubuntu Touch was not designed by me :) In fact, it was not designed at all but was instead a component developed ad-hoc.

**Comment by jdstrand** [comment]
@vosst (cc @ejratl) - I was referring to the storage framework for which I saw some user stories and designs a while back which I thought you did but I'm not up to date on the current implementation. Forgive me if I got the details wrong-- just trying to make sure that we are considering convergence here.

**Comment by vosst** [comment]
@jdstrand very much appreciated :) the scope of the storage framework does not include auto-mounting of removable storage volumes, yet. With that, we should probably distinguish between short term and long term solutions.
